src/grids_general.jl

```diff
@@ -0,0 +1,95 @@
+"""
+    abstract type AbstractGrid{T} <: AbstractVector{T} end
+
+The abstract supertype for all spatial grids on the sphere supported by SpeedyWeather.jl."""
+abstract type AbstractGrid{T} <: AbstractVector{T} end
+
+# all AbstractGrids have their grid points stored in a vector field `data`
+# propagate length, size, getindex, setindex! for that
+Base.length(G::AbstractGrid) = length(G.data)
+Base.size(G::AbstractGrid) = size(G.data)
+@inline function Base.getindex(G::AbstractGrid,k::Integer)
+    @boundscheck 0 < k <= length(G.data) || throw(BoundsError(G,k))
+    @inbounds r = G.data[k]
+    return r
+end
+
+@inline Base.setindex!(G::AbstractGrid,x,k::Integer) = setindex!(G.data,x,k)
+
+# with ranges
+@inline Base.getindex(G::AbstractGrid,r::AbstractRange) = G.data[r]
+@inline Base.setindex!(G::AbstractGrid,x::AbstractVector,r::AbstractRange) = setindex!(G.data,x,r)
+
+
+# List all the methods required to be implemented for any Grid types here
+# and define the  default behaviour as raising errors
+# This would clarify what interface we need to implement when introducing a new Grid
+
+truncation_order(::Type{G}) where {G<:AbstractGrid} =
+    error("truncation_order(::Type{G}) not implemented for G==$G")
```

Comment:
I'm not sure I like this. I see where you are coming from, but then I'd suggest don't `error("...not implemented")` but use a `throw(MethodError(f,args))` for that (because that's the type of error that's supposed to do that) but then that is exactly what happens anyway if you don't define the method in the first place. So I have another suggestion, which just writes all of that imformation explicitly down as a docstring """....""", that way one can simply do `?AbstractGrid` and gets that information printed into the REPL and it's also automatically included in the documentation.

```julia
"""
    abstract type AbstractGrid{T} <: AbstractVector end

The abstract supertype for all spatial grids on the sphere supported by SpeedyWeather.jl. Every new grid has to be of the form

    abstract type AbstractGridClass{T} <: AbstractGrid{T} end

    struct MyNewGrid{T} <: AbstractGridClass{T}
         data::Vector{T}     # all grid points unravelled into a vector
         nresolution::Int     # resolution parameter, like `nlat_half`, `nside` etc
    end

`MyNewGrid` should belong to a grid class like `AbstractFullGrid`, `AbstractOctahedralGrid` or
`AbstractHEALPixGrid` (that already exist but you may introduce a new class of grids) that share
certain features such as the number of longitude points per latitude ring and indexing, but may
have different latitudes or offset rotations. Each new grid `Grid` (or grid class) then has to
implement the following methods

Fundamental grid properties
    get_nresolution(::Grid)    # returns the resolution parameter
    get_npoints(::Grid)          # total number of grid points
    nlat_odd
    get_nlat
    get_nlat_half
    
Indexing
    get_nlon_max
    get_nlon_per_ring
    each_index_in_ring
    
Coordinates
    get_colat
    get_colatlon

Spectral truncation
    get_truncation
    get_resolution

Quadrature weights and solid angles
    get_quadrature_weights
    get_solid_angle
"""
AbstractGrid
```

Something like this?

Reply:
Thank you, Milan! Totally agreed. What you suggest is much more elegant and I like it. Thanks!

Reply:
Now in [50d47af](https://github.com/milankl/SpeedyWeather.jl/pull/138/commits/50d47af775fca6594085d9bc890409b4aecfcbb1)